### PR TITLE
fix(Core/Battleground): Improving the Anti-Spam BG Queue Announcer

### PR DESCRIPTION
**Pull request by wizzymore**
##### CHANGES PROPOSED:
-  Improving the AntiSpam BG Queue Announcer feature

##### ISSUES ADDRESSED:
- Closes https://github.com/azerothcore/azerothcore-wotlk/issues/2220

##### TESTS PERFORMED:
- Spamming the queue for a minute long on multiple accounts

##### HOW TO TEST THE CHANGES:

Spam the queue on multiple accounts for some time
##### Target branch(es):
- [x] Master

 

 
## How to test AzerothCore PRs
 
When a PR is ready to be tested, it will be marked as **[WAITING TO BE TESTED]**.

You can help by testing PRs and writing your feedback here in the PR's page on GitHub. Follow the instructions here:

http://www.azerothcore.org/wiki/How-to-test-a-PR

### CODE REVIEW

**Comment by pklloveyou** [comment]
Tested, perfect.

**Comment by Winfidonarleyan** [comment]
Please replace this
```cpp
struct BGSpamProtectionS
{
    uint32 last_queue = 0; // CHAT DISABLED BY DEFAULT
};

std::unordered_map<uint32, BGSpamProtectionS> BGSpamProtection;
```
maybe to 
```cpp
std::unordered_map<uint32, uint32> BGSpamProtection;
```

**Comment by wizzymore** [comment]
> Please replace this
> 
> ```c++
> struct BGSpamProtectionS
> {
>     uint32 last_queue = 0; // CHAT DISABLED BY DEFAULT
> };
> 
> std::unordered_map<uint32, BGSpamProtectionS> BGSpamProtection;
> ```
> 
> maybe to
> 
> ```c++
> std::unordered_map<uint32, uint32> BGSpamProtection;
> ```

Hi,

Why should i replace the struct with uint32? It will be very confusing.
I updated the first uint32 tho. I just found out that GUID can be bigger than int32

**Comment by wizzymore** [comment]
> Please replace this
> 
> ```c++
> struct BGSpamProtectionS
> {
>     uint32 last_queue = 0; // CHAT DISABLED BY DEFAULT
> };
> 
> std::unordered_map<uint32, BGSpamProtectionS> BGSpamProtection;
> ```
> 
> maybe to
> 
> ```c++
> std::unordered_map<uint32, uint32> BGSpamProtection;
> ```

I just got the point, i will replate it to uint32, you're right

**Comment by wizzymore** [comment]
Hi @pklloveyou can you please re-test this? I am unable to do so for the moment as i will have no access on the pc over the weekend.
For the linux part Travis will cover me

**Comment by pklloveyou** [comment]
Tested and working as expected

**Comment by FrancescoBorzi** [comment]
good job @wizzymore and thanks  @Winfidonarleyan @pklloveyou

**Comment by wizzymore** [comment]
> @wizzymore
> Could you please take a look at this?
> https://github.com/azerothcore/azerothcore-wotlk/blob/f39dac19840cdfa2260f4a9242d636b1013b198a/src/server/game/Battlegrounds/BattlegroundQueue.cpp#L1029
> 
> What happens if the GUID is not yet contained in the map? Wouldn't this cause an error?

How could he not be in a map? He need to trigger the action to get the message, you can only trigger the SendMessageQueue only when you are in a map, the leader will always be in the map it doesn't take in count the other party members if leader is in a party so why to check it?

**Comment by Stoabrogga** [comment]
@wizzymore 
I meant the unordered_map "BGSpamProtection". When you access the "leader->GetGUID()" for the first time there won't be a key in the unordered_map.

**Comment by wizzymore** [comment]
When adding a new key the map should be initialize with 0 automatically unless you give a custom map to the key

**Comment by wizzymore** [comment]
How i did it first time was okay using the struct but as you said, the methods are the same. And even tho the server will not initialize the unordered map with 0 ( even tho impossible ) it is 1 in a million chance that the result will be <=30 and only for a short time anyway. We still need to have another condition in case if we have strange data in the unordered map, checking if the difference is lower than 0 as only when the unordered map is randomly initialized can have values bigger than world time

**Comment by Stoabrogga** [comment]
Sorry, but this is also wrong. A uint32 can never be negative. Please use this:
```cpp
    else if (!bg->isArena()) // Show queue status to server (when joining battleground queue)
    {
        auto searchGUID = BGSpamProtection.find(leader->GetGUID());

        if (searchGUID == BGSpamProtection.end())
            BGSpamProtection[leader->GetGUID()] = 0; // Leader GUID not found, initialize with 0

        if (sWorld->GetGameTime() - BGSpamProtection[leader->GetGUID()] >= 30)
        {
            BGSpamProtection[leader->GetGUID()] = sWorld->GetGameTime();
            sWorld->SendWorldText(LANG_BG_QUEUE_ANNOUNCE_WORLD, bgName, q_min_level, q_max_level, qAlliance + qHorde, MaxPlayers);
        }
    }
```
This way it is checked if the key exists and if not it inserts a new one and initializes the value with 0.

**Comment by wizzymore** [comment]
Ffs dude i will revert all the changes

When you have an idea of what you are saying then create the god damn change request in the first place cause you are already annoying

**Comment by wizzymore** [comment]
I did the changes, sorry again for that comment.

I will test this PR tonight

**Comment by Stoabrogga** [comment]
> sorry again for that comment.

No offence taken. I think we just talked past each other.

**Comment by FrancescoBorzi** [comment]
excellent job guys, well done!

File: src/server/game/Battlegrounds/BattlegroundQueue.cpp [hunk 1]
```diff
@@ -1031,17 +1026,15 @@ void BattlegroundQueue::SendMessageQueue(Player* leader, Battleground* bg, PvPDi
...
+        if (searchGUID == BGSpamProtection.end())
+            BGSpamProtection[leader->GetGUID()] = 0; // Leader GUID not found, initialize with 0
+
+        if (sWorld->GetGameTime() - BGSpamProtection[leader->GetGUID()] >= 30)
         {
-            BGSpamProtection[leader->GetGUID()].last_queue = 0;
-            sWorld->SendWorldText(LANG_BG_QUEUE_ANNOUNCE_WORLD, bgName, q_min_level, q_max_level,
-                qAlliance + qHorde, MaxPlayers);
+            BGSpamProtection[leader->GetGUID()] = sWorld->GetGameTime();
+            sWorld->SendWorldText(LANG_BG_QUEUE_ANNOUNCE_WORLD, bgName, q_min_level, q_max_level, qAlliance + qHorde, MaxPlayers);
```
Comment (pklloveyou) [inline src/server/game/Battlegrounds/BattlegroundQueue.cpp]:
```c++
					else
						ChatHandler(leader->GetSession()).PSendSysMessage(LANG_BG_QUEUE_ANNOUNCE_WORLD, bgName, q_min_level, q_max_level, qAlliance + qHorde, MaxPlayers);
```

File: src/server/game/Battlegrounds/BattlegroundQueue.cpp [hunk 2]
```diff
@@ -1031,17 +1026,15 @@ void BattlegroundQueue::SendMessageQueue(Player* leader, Battleground* bg, PvPDi
...
         }
```
Comment (pklloveyou) [inline src/server/game/Battlegrounds/BattlegroundQueue.cpp]:
```c++
else
  ChatHandler(leader->GetSession()).PSendSysMessage(LANG_BG_QUEUE_ANNOUNCE_WORLD, bgName, q_min_level, q_max_level, qAlliance + qHorde, MaxPlayers);
```

Comment (pklloveyou) [reply src/server/game/Battlegrounds/BattlegroundQueue.cpp]: What if it does?

**Pull request merged**